controllers/db/bytebase/controllers/bytebase_controller.go

```diff
@@ -0,0 +1,226 @@
+//+kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=networking.k8s.io,resources=ingresses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=rolebindings,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apisix.apache.org,resources=apisixroutes,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apisix.apache.org,resources=apisixtlses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=acid.zalan.do,resources=postgresqls,verbs=get;list;watch;create;update;patch;delete
+
+func (r *BytebaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger
+	bb := &bbv1.Bytebase{}
+	// Get CRD bytebase (status). Call reconciler again if not found.
+	if err := r.Get(ctx, req.NamespacedName, bb); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if bb.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(bb, FinalizerName) {
+			controllerutil.AddFinalizer(bb, FinalizerName)
+			if err := r.Update(ctx, bb); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		logger.Info("remove finalizer")
+		if controllerutil.ContainsFinalizer(bb, FinalizerName) {
+			r.Bc = nil
+			controllerutil.RemoveFinalizer(bb, FinalizerName)
+			if err := r.Update(ctx, bb); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.fillDefaultValue(ctx, bb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if isExpired(bb) {
+		if err := r.Delete(ctx, bb); err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Info("successfully deleted the expired bytebase resource.")
+		return ctrl.Result{}, nil
+	}
+
+	var hostname string
+	if err := r.syncDeployment(ctx, bb, &hostname); err != nil {
+		logger.Error(err, "create deployment failed")
+		r.Recorder.Eventf(bb, corev1.EventTypeWarning, "create deployment failed", "%v", err)
+		return ctrl.Result{}, err
+	}
+
+	if err := r.syncService(ctx, bb); err != nil {
+		logger.Error(err, "create service failed")
+		r.Recorder.Eventf(bb, corev1.EventTypeWarning, "create service failed", "%v", err)
+		return ctrl.Result{}, err
+	}
+
+	// Init instance logic here
+	/// get basic info, bytebase entry point etc
+	svc := corev1.Service{}
+	if err := r.Get(ctx, client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      bb.Name,
+	}, &svc); err != nil {
+		logger.Error(err, "get bytebase service failed")
+		return ctrl.Result{}, err
+	}
+
+	url := "http://" + svc.Spec.ClusterIP + ":" + bb.Spec.Port.String()
```

Comment:
Why not use domain name?  fmt.Sprintf("http://%s:%s",,) is better.

Reply:
Users can change the domain name of the cluster through configuring kubelet, domain name like `*.cluster.local` is not always true for the cluster. We might need to start a conversation with kubelet to know the actual domain name(Is there a way to use kubernetes api (api-server) to achieve the same thing aside from open a proxy?). I think we should keep what api-server has to offer, i.e. the service IP.

Reply:
> fmt.Sprintf("http://%s:%s",,) is better.

This is resolved by latest commits.